dgc/GreenCertificateVerifier.cs

```diff
@@ -17,28 +18,30 @@ public GreenCertificateVerifier(SecretariatService secretariatService)
         /// </summary>
         /// <param name="coseMessage">Cose message</param>
         /// <returns>(true, null) if valid, (false, reason) if not valid</returns>
-        public (bool, string) Verify(CWT cwt)
+        public async Task<Tuple<bool, string>> Verify(CWT cwt, DateTime? verifyClock = null)
         {
-            var publicKeys = _secretariatService.GetPublicKeys(cwt.CoseMessage.KID);
-            if (publicKeys.Any())
+            if (!verifyClock.HasValue) verifyClock = DateTime.Now;
+
+            var certificates = await _secretariatService.GetCertificate(cwt.CoseMessage.KID);
+            if (certificates.Any())
```

Comment:
Multiple enumeration; usually I'd have the service return a collection type (i.e. readonly list) here, then you get a enumerator each time :)

Reply:
> Multiple enumeration; 

In what way? `Any()` just checks to see if there is anything in the collection, doesn't really enumerate in the sense of going through the entire collection: ["The enumeration of source is stopped as soon as the result can be determined."](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=net-5.0) - which really should just check if there is an initial element or not and not even try to enumerate. Given the disappointing code I've seen when using ILDASM (even recently) on C# / .NET 5x, I could well imagine it still at least starts to enumerate rather than just directly check for presence of first element.

That said, I don't see why you need the call to `Any()` as the `foreach()` would simply nop if the collection empty - but in any case, this discussion of calling `Any()` is - at least AFAICS - more an optimization point rather than a program correctness point

> usually I'd have the service return a collection type (i.e. readonly list) here, then you get a enumerator each time :)

isn't that what `GetCertificate()` does? (which should really be called `GetCertificates()` ) - or have I just misread / assumed something that isn't there?

Reply:
I should have said "potential multiple enumeration" - in the code base as it is now it is not a bug, but a code smell. 

The interface being called exposes IEnumerable<>, the contract that this defines is "enumerate this". This can lead to unexpected behaviour - in most cases when people expose IEnumerable directly via an ORM (and something that you - sadly - see way to often).

Despite never having encountered the bug in my own code (I know what I'm doing), I've seen it so often that I my code analysis tools are configured to treat this as an actual error :-)

Reply:
> I should have said "potential multiple enumeration" - in the code base as it is now it is not a bug, but a code smell.
> 

? If the .net framework internally implements `Any()` as an enumeration process returning a boolean or does sth more sensible is nothing the caller of the framework method `Any()` can decide. As a caller, you can of course choose not to call `Any()` ;) Any code smell would therefore also be in the .NET framework rather than end-user code.

> The interface being called

well you would be calling an implementation of an interface, an interface by definition is not directly callable (abstract). So you would be calling something that implements the interface IEnumerable, in this case a Collection class.

> exposes IEnumerable<>, the contract that this defines is "enumerate this".

Here the return value is an object of type "Collection" (unless I'm mis-reading the code?) which amongst other things implements the interface `IEnumerable`, but [IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-5.0) itself does not have an 'Any()' method. The 'Any()' method is specified in [linq](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.any?view=net-5.0) and takes an `IEnumerable` argument, so yes the linq impl of `Any()` would be constrained to calling one of `IEnumerable.Current / MoveNext / Reset` but as the documentation for `Any()` states (ref in earlier comment) `Any()` will return as soon as it can determine the return value, so at least it doesn't run through the whole collection here, which was my initial point.

> This can lead to unexpected behaviour - in most cases when people expose IEnumerable directly via an ORM (and something that you - sadly - see way to often).

... but I don't see that is the case here - or am I missing sth?

In any case, the IEnumerable implementation called here is exposed from the .NET framework implementation for the Collection classes so no user-defined code here either, just calls to framework. Same for the (possibly unnecessary - separate issue) call to `Any()`. Not unknown for the .NET framework to have bugs in it, but that's a separate issue again.

Reply:
Nope you're not missing anything - it's a "code cleanness" / pedantic code analysis point (so nothing will go wrong in the current implementation):

The interface returns an IEnumerable<X509Certificate2>:

`Task<IEnumerable<X509Certificate2>> GetCertificate(string kid);`

It's like declaring an interface that can return `Fruit`, when you really only want to see `CitrusFruit` with the default implementations doing just that. Fine, it will work, at least until some clever person adds a new implementation that returns `Apples`.

So, in this case, there's not a *need* to change anything, but doing so does improve the code.

Reply:
> It's like declaring an interface that can return Fruit, when you really only want to see CitrusFruit

kind-of, but a little different: it's the generic type binding to IEnumerable (`X509Certificate2`) which makes it `CitrusFruit` -> but then here you have a very specific type so returning e.g. `IEnumerable<MyOtherType>` should be caught by the compiler (unless there is some clear inheritance relationship between `X509Certificate2` and `MyOtherType` - LSP, in effect).

Reply:
I changed this to IReadOnlyList